### PR TITLE
TYP: fix a few types

### PR DESCRIPTION
**Pull request by twoertwein**
I fixed a few type issues discovered by using the pandas-stubs type tests with the pandas annotations (there are many errors).

I'm surprised that the two Excel-classes being marked as abstract wasn't detected by the CI, they are clearly meant to be instantiated, so they cannot be abstract.

### CODE REVIEW

File: pandas/__init__.py [hunk 1]
```diff
@@ -110,6 +110,7 @@
 
 from pandas.core.dtypes.dtypes import SparseDtype
 
+from pandas import util
```
Comment (twoertwein) [inline pandas/__init__.py:113]: For https://pandas.pydata.org/docs/reference/api/pandas.util.hash_pandas_object.html and other functions

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1926,11 +1926,17 @@ def to_dict(
         self,
         orient: Literal["dict", "list", "series", "split", "tight", "index"] = ...,
         into: type[dict] = ...,
+        index: bool = ...,
```
Comment (twoertwein) [inline pandas/core/frame.py:1929]: Present in the implementation but missing in the overloads

File: pandas/core/frame.py [hunk 3]
```diff
@@ -11711,6 +11717,7 @@ def quantile(
         axis: Axis = ...,
         numeric_only: bool = ...,
         interpolation: QuantileInterpolation = ...,
+        method: Literal["single", "table"] = ...,
```
Comment (twoertwein) [inline pandas/core/frame.py:11720]: Present in the implementation but missing in the overloads

File: pandas/__init__.py (hunk 1 above)
Comment (mroeschke) [reply pandas/__init__.py:113]: I think there are other methods in this module that are not meant to be public though: `capitalize_first_letter`, `cache_readonly`, `Substitution`, `Appender`

Comment (twoertwein) [reply pandas/__init__.py:113]: Okay, I think we could do `__all__ = [all_classes_functions_in_the_docs]` to account for that. And/or we could deprecate it from utils and put it as part of a different namespace.

File: pandas/io/parsers/readers.py [hunk 4]
```diff
@@ -1374,6 +1421,9 @@ def read_fwf(
     --------
     >>> pd.read_fwf('data.csv')  # doctest: +SKIP
     """
+    kwds["iterator"] = iterator
+    kwds["chunksize"] = chunksize
```
Comment (rhshadrach) [inline pandas/io/parsers/readers.py:1425]: Nit: Would it be better to do this down on L1462 with the other assignments?

File: pandas/__init__.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/__init__.py:113]: `__all__` only changes behavior of `from foo import *`, no? What do we think of moving these functions (can we make a list of them?) to the top namespace?

Does this hold up other aspects of this PR? If not, I would suggest moving to a separate PR.

Comment (twoertwein) [reply pandas/__init__.py:113]: Happy to remove this change from this PR. I will open an issue to first discuss how this should be handled.

`__all__` affects `from pandas.utils import *` but (more importantly) type checkers, use it (next to other rules) to determine which symbols are meant to be public: if `__all__` is present, only those symbols listed in it are considered public.

Comment (rhshadrach) [reply pandas/__init__.py]: > if __all__ is present, only those symbols listed in it are considered public.

This doesn't seem to be the case with mypy.

```
# foo.py
x = 5
y = 6

__all__ = ['y']

# bar.py
import foo

print(foo.x)
print(foo.y)
```

The result of `mypy bar.py`: `Success: no issues found in 1 source file`

Comment (rhshadrach) [reply pandas/__init__.py]: > if `__all__` is present, only those symbols listed in it are considered public.

This doesn't seem to be the case with mypy:

```
# test/__init__.py
x = 5
y = 6

__all__ = ['y']

# bar.py
import test

print(test.x)
print(test.y)
```

`mypy bar.py` reports no issues for me.

Comment (twoertwein) [reply pandas/__init__.py]: I must have misremembered that (pyright also allows that).

Pyright describes how it handles it: https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface

File: pandas/io/formats/excel.py [hunk 5]
```diff
@@ -941,9 +941,7 @@ def write(
         if isinstance(writer, ExcelWriter):
             need_save = False
         else:
-            # error: Cannot instantiate abstract class 'ExcelWriter' with abstract
-            # attributes 'engine', 'save', 'supported_extensions' and 'write_cells'
-            writer = ExcelWriter(  # type: ignore[abstract]
+            writer = ExcelWriter(
```
Comment (rhshadrach) [inline pandas/io/formats/excel.py:944]: Won't this immediately raise in the `try` below?

Comment (rhshadrach) [reply pandas/io/formats/excel.py:944]: Ahh, no, because we overwrite `__new__`. I think this is okay.

**Review by rhshadrach — APPROVED** [review]
lgtm; just leaving a few comments in case the bit that confused me gets anyone else.

**Comment by rhshadrach** [comment]
@twoertwein - looks like isort is making a modification

**Comment by twoertwein** [comment]
> @twoertwein - looks like isort is making a modification

All green now :)

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @twoertwein

**Pull request merged**